aijson/utils/action_utils.py

```diff
@@ -260,82 +262,112 @@ def build_action_description(
 
 
 def build_link_literal(
-    action_invocations: dict[ExecutableId, ActionInvocation],
+    config_filename: str,
     strict: bool,
     include_paths: bool,
-) -> type[str]:
-    union_elements = []
+) -> tuple[type[str], dict[str, UnionType | None]]:
+    from aijson.models.config.flow import ActionConfig, Loop
+    from aijson.utils.loader_utils import load_config_file
+
+    try:
+        # load the file not as a non-strict model
+        action_config = load_config_file(config_filename, config_model=ActionConfig)
+    except pydantic.ValidationError:
+        log = get_logger()
+        log.debug(
+            "Failed to load action config",
+            exc_info=True,
+        )
+        return (str, {})
+
+    # actions = get_actions_dict()
 
+    union_elements = []
     if not strict:
         union_elements.append(str)
 
     actions_dict = get_actions_dict()
+    loop_elements: dict[str, UnionType | None] = {}
+
+    def build(action_config: ActionConfig):
+        union_elements = []
+        for action_id, action_invocation in action_config.flow.items():
+            if isinstance(action_invocation, (Loop)):
+                # TODO fixing Loop
+                union_elements.append(Literal[action_id])
+                links = build(action_invocation)  # type: ignore
+                loop_elements[action_id] = links
+                continue
+            if isinstance(action_invocation, (ValueDeclaration)):
+                union_elements.append(Literal[action_id])
+                continue
```

Comment:
can you make these check all possible types?
eg

```python
if isinstance(action_invocation, Loop):
    ...
elif isinstance(action_invocation, ValueDeclaration):
    ...
elif isinstance(action_invocation, ActionInvocation):
    ...
else:
    assert_never(action_invocation)
```

this will make typing complain whenever we add a new executable type

also, rename `action_invocation` to `executable_invocation`, or something more general